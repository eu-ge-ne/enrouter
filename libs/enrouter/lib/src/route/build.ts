import * as regexparam from "regexparam";

import { logger } from "#lib/debug.js";

import type { Route } from "./mod.js";

const log = logger("route/build");

export interface BuildRoutesParams {
  modules: RouteModules;
}

/**
 * Builds `Route`s from `RouteModules`
 */
export function buildRoutes({ modules }: BuildRoutesParams): Route | undefined {
  log("Building routes");

  const entries = Object.entries(modules).sort(
    (a, b) => a[1].dirPath.length - b[1].dirPath.length,
  );

  const routes = new Map<string, Route>();

  function findParent(dp: string[]) {
    const parentPath = [...dp];

    while (parentPath.length > 0) {
      parentPath.pop();
      const path = parsePath("/" + parentPath.join("/"));
      const parent = routes.get(path);
      if (parent) {
        return parent;
      }
    }

    throw new Error("Parent not found");
  }

  for (const [moduleId, { dirPath, fileName, load }] of entries) {
    const isRoot = dirPath.length === 0;

    const path = isRoot ? "/" : parsePath("/" + dirPath.join("/"));

    let route = routes.get(path);
    if (!route) {
      route = {
        path,
        test: regexparam.parse(path, true),
        modules: [],
        loaded: false,
        elements: {},
      };
      routes.set(path, route);
    }

    if (!isRoot) {
      const parent = findParent(dirPath);
      if (!parent.tree?.find((x) => x.path === path)) {
        if (!parent.tree) {
          parent.tree = [];
        }
        parent.tree.push(route);
      }
    }

    route.modules.push({
      id: moduleId,
      fileName,
      load,
    });
  }

  const result = routes.get("/");

  log("Routes built: %o", routes);

  return result;
}

function parsePath(str: string) {
  return str.replace(/\[(.+)\]/, ":$1");
}

/**
 * Collection of route module descriptors.
 * Maps module id to fs path and async import function.
 */
type RouteModules = Record<
  string,
  {
    dirPath: string[];

    fileName: string;
    /**
     * Async module import function
     */
    load: () => Promise<unknown>;
  }
>;
