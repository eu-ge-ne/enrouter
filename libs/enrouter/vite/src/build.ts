import * as regexparam from "regexparam";

import { logger } from "#lib/debug.js";

import type { Route } from "#lib/route/mod.js";

const log = logger("route/build");

export type RouteModules = {
  id: string;
  dir: string[];
  fileName: string;
  load: () => Promise<unknown>;
}[];

/**
 * Builds `Route`s from `RouteModules`
 */
export function buildRoutes(modules: RouteModules): Route | undefined {
  log("Building routes");

  const sorted = modules.sort((a, b) => a.dir.length - b.dir.length);

  const routes = new Map<string, Route>();

  function findParent(dp: string[]) {
    const parentPath = [...dp];

    while (parentPath.length > 0) {
      parentPath.pop();
      const path = parsePath("/" + parentPath.join("/"));
      const parent = routes.get(path);
      if (parent) {
        return parent;
      }
    }

    throw new Error("Parent not found");
  }

  for (const { id, dir, fileName, load } of sorted) {
    const isRoot = dir.length === 0;

    const path = isRoot ? "/" : parsePath("/" + dir.join("/"));

    let route = routes.get(path);
    if (!route) {
      route = {
        path,
        test: regexparam.parse(path, true),
        modules: [],
        loaded: false,
        elements: {},
      };
      routes.set(path, route);
    }

    if (!isRoot) {
      const parent = findParent(dir);
      if (!parent.tree?.find((x) => x.path === path)) {
        if (!parent.tree) {
          parent.tree = [];
        }
        parent.tree.push(route);
      }
    }

    route.modules.push({
      id,
      fileName,
      load,
    });
  }

  const result = routes.get("/");

  log("Routes built: %o", routes);

  return result;
}

function parsePath(str: string) {
  return str.replace(/\[(.+)\]/, ":$1");
}

export function routeToJS(
  route: Route,
  getLoad: (id: string) => string,
  tab = 0,
): string {
  const tree = route.tree
    ? route.tree.map((x) => routeToJS(x, getLoad, 4))
    : undefined;
  const treeStr = tree
    ? `[
${tree.join(",\n")}
  ]`
    : undefined;

  const mods = route.modules
    .map(
      (x) => `
{
    id: "${x.id}",
    fileName: "${x.fileName}",
    load: ${getLoad(x.id)},
}`,
    )
    .join(",")
    .replace(/^/gm, " ".repeat(4));

  return `{
    path: "${route.path}",
    test: {
      keys: [${route.test.keys.map((x) => "${x}").join(",")}],
      pattern: ${route.test.pattern},
    },
    modules: [${mods}],
    loaded: false,
    elements: {},
    tree: ${treeStr},
}`.replace(/^/gm, " ".repeat(tab));
}
