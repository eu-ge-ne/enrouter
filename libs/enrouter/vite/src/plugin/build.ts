import * as regexparam from "regexparam";

import type { Route } from "#lib/route/mod.js";
import type { RouteModules } from "./modules.js";

/**
 * Builds `Route`s from `RouteModules`
 */
export function buildRoutes(modules: RouteModules): Route {
  const sorted = modules.sort((a, b) => a.dir.length - b.dir.length);

  const routes = new Map<string, Route>();

  function findParent(dp: string[]) {
    const parentPath = [...dp];

    while (parentPath.length > 0) {
      parentPath.pop();
      const path = parsePath("/" + parentPath.join("/"));
      const parent = routes.get(path);
      if (parent) {
        return parent;
      }
    }

    throw new Error("Parent not found");
  }

  for (const { dir, isRoot, id, fileName, importFn } of sorted) {
    const path = isRoot ? "/" : parsePath("/" + dir.join("/"));

    let route = routes.get(path);
    if (!route) {
      route = {
        path,
        test: regexparam.parse(path, true),
        modules: [],
        loaded: false,
        elements: {},
      };
      routes.set(path, route);
    }

    if (!isRoot) {
      const parent = findParent(dir);
      if (!parent.tree?.find((x) => x.path === path)) {
        if (!parent.tree) {
          parent.tree = [];
        }
        parent.tree.push(route);
      }
    }

    route.modules.push({
      id,
      fileName,
      importFn,
    });
  }

  const root = routes.get("/");
  if (!root) {
    throw new Error("Routes were not built");
  }

  return root;
}

function parsePath(str: string) {
  return str.replace(/\[(.+)\]/, ":$1");
}
