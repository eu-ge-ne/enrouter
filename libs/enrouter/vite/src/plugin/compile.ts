import type { Route } from "#lib/route/mod.js";
import type { RouteModules } from "./modules.js";

export function compileRoutes(
  modules: RouteModules[],
  route: Route,
  tab = 0,
): string {
  const tree = route.tree
    ? route.tree.map((x) => compileRoutes(modules, x, 4))
    : undefined;

  const treeStr = tree
    ? `[
${tree.join(",\n")}
  ]`
    : undefined;

  const mods = route.modules
    .map(
      (x) => `
{
    id: "${x.id}",
    fileName: "${x.fileName}",
    importFn: ${modules.find(({ routePath }) => routePath === route.path)!.routeModules.find(({ id }) => id === x.id)?.importStr},
}`,
    )
    .join(",")
    .replace(/^/gm, " ".repeat(4));

  return `{
    path: "${route.path}",
    test: {
      keys: [${route.test.keys.map((x) => "${x}").join(",")}],
      pattern: ${route.test.pattern},
    },
    modules: [${mods}],
    loaded: false,
    elements: {},
    tree: ${treeStr},
}`.replace(/^/gm, " ".repeat(tab));
}
