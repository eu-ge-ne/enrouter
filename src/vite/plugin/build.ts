import { parseRoutePath } from "./modules.js";
import type { Route } from "#lib/route/mod.js";
import type { RouteModules } from "./modules.js";

export function buildRouteTree(modules: RouteModules[]): Route {
  const routes = new Map<string, Route>();

  function findParent(dp: string[]): Route | undefined {
    if (dp.length === 0) {
      return;
    }

    const parentPath = [...dp];

    while (parentPath.length > 0) {
      parentPath.pop();
      const { path } = parseRoutePath(parentPath);
      const parent = routes.get(path);
      if (parent) {
        return parent;
      }
    }

    throw new Error("Parent not found");
  }

  const sorted = modules.sort((a, b) => a.dir.length - b.dir.length);

  for (const { dir, path, test, modules } of sorted) {
    let route = routes.get(path);
    if (!route) {
      route = {
        path,
        test,
        modules: modules.map(({ importStr, ...x }) => x),
        loaded: false,
        elements: {},
      };
      routes.set(path, route);
    }

    const parent = findParent(dir);
    if (parent) {
      parent.tree = parent.tree ?? [];
      parent.tree.push(route);
    }
  }

  const root = routes.get("/");
  if (!root) {
    throw new Error("Routes were not built");
  }

  return root;
}
