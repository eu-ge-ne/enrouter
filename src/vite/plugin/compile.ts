import { parseRoutePath } from "./modules.js";

import type { RouteModules } from "./modules.js";

interface RouteProps {
  props: string;
  tree?: RouteProps[];
}

export function compileRouteTree(modules: RouteModules[]): string {
  const routes = new Map<string, RouteProps>();

  function findParent(dp: string[]): RouteProps | undefined {
    if (dp.length === 0) {
      return;
    }

    const parentPath = [...dp];

    while (parentPath.length > 0) {
      parentPath.pop();
      const { path } = parseRoutePath(parentPath);
      const parent = routes.get(path);
      if (parent) {
        return parent;
      }
    }

    throw new Error("Parent not found");
  }

  const sorted = modules.sort((a, b) => a.dir.length - b.dir.length);

  for (const { dir, path, test, modules } of sorted) {
    let route = routes.get(path);
    if (!route) {
      const mods = modules
        .map((x) =>
          `{
  id: "${x.id}",
  fileName: "${x.fileName}",
  importFn: ${x.importStr},
}`.replace(/^/gm, " ".repeat(2)),
        )
        .join(",\n");

      route = {
        props: `path: "${path}",
test: {
  keys: [${test.keys.map((x) => "${x}").join(",")}],
  pattern: ${test.pattern},
},
modules: [
${mods}
],
loaded: false,
elements: {},`.replace(/^/gm, " ".repeat(2)),
      };
      routes.set(path, route);
    }

    const parent = findParent(dir);
    if (parent) {
      parent.tree = parent.tree ?? [];
      parent.tree.push(route);
    }
  }

  const root = routes.get("/");
  if (!root) {
    throw new Error("Routes were not compiled");
  }

  return `export default ${compileProps(root)};`;
}

function compileProps(route: RouteProps, tab = 0): string {
  const tree = route.tree
    ? `
tree: [
${route.tree.map((x) => compileProps(x, 2)).join(",\n")}
]`.replace(/^/gm, " ".repeat(2))
    : "";

  const res = `{
${route.props}${tree}
}`;

  return res.replace(/^/gm, " ".repeat(tab));
}
