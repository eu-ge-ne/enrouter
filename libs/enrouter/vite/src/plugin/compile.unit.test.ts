import * as vm from "node:vm";

import { describe, test, expect } from "vitest";
import * as regexparam from "regexparam";

import type { Route } from "#lib/route/mod.js";
import type { RouteModules } from "./modules.js";
import { buildRouteTree } from "./build.js";
import { compileRouteTree } from "./compile.js";

describe("compileRouteTree", () => {
  test("from 0 modules", () => {
    expect(() => compileRouteTree([])).toThrowErrorMatchingSnapshot();
  });

  test("from 1 module", async () => {
    const modules: RouteModules[] = [
      {
        dir: [],
        path: "/",
        test: regexparam.parse("/", true),
        modules: [
          {
            id: "src/_page.tsx",
            fileName: "_page.tsx",
            importFn: async () => {},
            importStr: '() => import("/home/src/_page.tsx")',
          },
        ],
      },
    ];

    const builtTree = buildRouteTree(modules);

    const compiled = compileRouteTree(modules);
    const compiledTree = await getCompiledTree(compiled);

    deleteImports(builtTree);
    deleteImports(compiledTree);

    expect(builtTree).toEqual(compiledTree);
  });

  test("from 2 modules", async () => {
    const modules: RouteModules[] = [
      {
        dir: [],
        path: "/",
        test: regexparam.parse("/", true),
        modules: [
          {
            id: "src/_page.tsx",
            fileName: "_page.tsx",
            importFn: async () => {},
            importStr: '() => import("/home/src/_page.tsx")',
          },
        ],
      },
      {
        dir: ["abc"],
        path: "/abc",
        test: regexparam.parse("/abc", true),
        modules: [
          {
            id: "src/abc/_page.tsx",
            fileName: "_page.tsx",
            importFn: async () => {},
            importStr: '() => import("/home/src/abc/_page.tsx")',
          },
        ],
      },
    ];

    const builtTree = buildRouteTree(modules);

    const compiled = compileRouteTree(modules);
    const compiledTree = await getCompiledTree(compiled);

    deleteImports(builtTree);
    deleteImports(compiledTree);

    expect(builtTree).toEqual(compiledTree);
  });

  test("from 3 modules", async () => {
    const modules: RouteModules[] = [
      {
        dir: [],
        path: "/",
        test: regexparam.parse("/", true),
        modules: [
          {
            id: "src/_page.tsx",
            fileName: "_page.tsx",
            importFn: async () => {},
            importStr: '() => import("/home/src/_page.tsx")',
          },
        ],
      },
      {
        dir: ["xyz"],
        path: "/xyz",
        test: regexparam.parse("/xyz", true),
        modules: [
          {
            id: "src/xyz/_page.tsx",
            fileName: "_page.tsx",
            importFn: async () => {},
            importStr: '() => import("/home/src/xyz/_page.tsx")',
          },
          {
            id: "src/xyz/_index.tsx",
            fileName: "_index.tsx",
            importFn: async () => {},
            importStr: '() => import("/home/src/xyz/_index.tsx")',
          },
        ],
      },
    ];

    const builtTree = buildRouteTree(modules);

    const compiled = compileRouteTree(modules);
    const compiledTree = await getCompiledTree(compiled);

    deleteImports(builtTree);
    deleteImports(compiledTree);

    expect(builtTree).toEqual(compiledTree);
  });
});

async function getCompiledTree(compiled: string): Promise<Route> {
  const script = new vm.SourceTextModule(compiled, {
    context: vm.createContext({}),
  });
  //@ts-ignore
  await script.link(() => {});
  await script.evaluate();
  //@ts-ignore
  return script.namespace.default;
}

function deleteImports(route: Route) {
  for (const mod of route.modules) {
    //@ts-ignore
    delete mod.importFn;
  }
  if (route.tree) {
    for (const child of route.tree) {
      deleteImports(child);
    }
  }
}
