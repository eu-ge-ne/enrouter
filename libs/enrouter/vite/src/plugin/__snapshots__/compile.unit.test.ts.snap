// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compileRouteTree > from 0 modules 1`] = `[Error: Routes were not compiled]`;
